gc.c

```diff
@@ -3131,13 +3131,11 @@ obj_free(rb_objspace_t *objspace, VALUE obj)
             rb_id_table_free(RCLASS_CVC_TBL(obj));
 	}
 	if (RCLASS_SUBCLASSES(obj)) {
-	    if (BUILTIN_TYPE(obj) == T_MODULE) {
-		rb_class_detach_module_subclasses(obj);
-	    }
-	    else {
-		rb_class_detach_subclasses(obj);
-	    }
-	    RCLASS_SUBCLASSES(obj) = NULL;
+            if (BUILTIN_TYPE(obj) == T_MODULE) {
+                rb_class_detach_module_subclasses(obj);
+            } else {
+                rb_class_remove_subclass_head(obj);
+            }
```

Comment:
This removes every node of the linked list for the superclass. The superclass should only be remove the dummy head.